src/attesters/sgx-ecdsa/collect_endorsements.c

```diff
@@ -0,0 +1,119 @@
+/* Copyright (c) 2021 Intel Corporation
+ * Copyright (c) 2020-2023 Alibaba Cloud
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <unistd.h>
+#include <rats-tls/log.h>
+#include <rats-tls/attester.h>
+#include <stddef.h>
+
+// clang-format off
+#ifdef SGX
+#include "sgx_lfence.h"
+#include "rtls_t.h"
+// clang-format on
+
+enclave_attester_err_t sgx_ecdsa_collect_endorsements(enclave_attester_ctx_t *ctx,
+						      attestation_evidence_t *evidence,
+						      attestation_endorsement_t *endorsements)
+{
+	enclave_attester_err_t ret = ENCLAVE_ATTESTER_ERR_NONE;
+	uint8_t *collateral_untrusted = NULL; /* address of collateral in untrusted-app */
+
+	RTLS_DEBUG("ctx %p, evidence %p, endorsements %p\n", ctx, evidence, endorsements);
+
+	int sgx_status = ocall_tee_qv_get_collateral(
+		&ret, evidence->ecdsa.quote, evidence->ecdsa.quote_len, &collateral_untrusted);
+	if (SGX_SUCCESS != sgx_status || ENCLAVE_ATTESTER_ERR_NONE != ret) {
+		RTLS_ERR("tee_qv_get_collateral(): 0x%04x, 0x%04x\n", sgx_status, ret);
+		if (SGX_SUCCESS != sgx_status)
+			ret = ENCLAVE_ATTESTER_ERR_UNKNOWN;
+		goto err;
+	}
+	RTLS_DEBUG("collateral_untrusted %p, collateral_size %p\n", ctx, collateral_untrusted);
+
+	/* Since we have to access memory pointed to by collateral_untrusted, before that we check if this memory is in the untrusted-app. */
+	if (!sgx_is_outside_enclave(collateral_untrusted, sizeof(sgx_ql_qve_collateral_t))) {
+		ret = ENCLAVE_ATTESTER_ERR_INVALID;
+		goto err;
+	}
+	/* fence after boundary check */
+	sgx_lfence();
+
+	/* Copy collateral struct from untrusted-app to enclave memory */
+	sgx_ql_qve_collateral_t collateral;
+	memcpy(&collateral, collateral_untrusted, sizeof(sgx_ql_qve_collateral_t));
+
+	/* Copy fields from collateral to endorsements->ecdsa */
+	sgx_ql_qve_collateral_t *c = &collateral;
```

Comment:
The use of  pointer `c` instead of `collateral` is for uniformity with pointer  `u`. This makes the macro _COPYENDORSEMENT_FIELD look more concise. :)